tests/unit/test__fixture__cmdln_args.py

```diff
@@ -0,0 +1,161 @@
+            ),
+        )
+    )
+
+    perform_fixture_test_with_optional_log_capture(pytester)
+
+
+def test__cmdln_args__empty_cmdln(pytester):
+    """Fixture should return an empty config if command line is empty."""
+    # note, we are _not_ using magic mock here because pytest's request.config
+    # object is very hard to patch. See the definition of MockConfig (conftest)
+    # for more details.
+    patches = """
+        from tests.unit.conftest import MockConfig
+
+        @pytest.fixture(scope="session")
+        def request_config():
+            mock = MockConfig()
+            return mock
+        """
+    pytester.makeconftest(
+        CONFTEST_TEMPLATE.format(fixture_name=FIXTURE_UNDER_TEST, patches=patches)
+    )
+    pytester.makepyfile(
+        PYFILE_TEMPLATE.format(
+            fixture_name=FIXTURE_UNDER_TEST,
+            validations=PYFILE_ASSERT_EMPTY_CONFIG.format(
+                fixture_name=FIXTURE_UNDER_TEST
+            ),
+        )
+    )
+
+    perform_fixture_test_with_optional_log_capture(pytester)
+
+
+def test__cmdln_args__extra_args_coming_from_cmdln(pytester):
+    """Fixture should load any attrs that match COMPLETE_ATTRS and ignore any other
+    attrs in the collection, such as pytest args or custom args that a consumer might
+    add in."""
+
+    patches = """
+        from tests.unit.conftest import MockConfig
+
+        @pytest.fixture(scope="session")
+        def request_config():
+            mock = MockConfig({"--users": 42, "--log_cli_level": "debug"})
+            return mock
+        """
+    pytester.makeconftest(
+        CONFTEST_TEMPLATE.format(fixture_name=FIXTURE_UNDER_TEST, patches=patches)
+    )
+    expected_config_args = {"users": 42}
+    pytester.makepyfile(
+        PYFILE_TEMPLATE.format(
+            fixture_name=FIXTURE_UNDER_TEST,
+            validations=PYFILE_ASSERT_EXPECTED_CONFIG.format(
+                fixture_name=FIXTURE_UNDER_TEST, expected=expected_config_args
+            ),
+        )
+    )
+
+    perform_fixture_test_with_optional_log_capture(pytester)
+
+
+@patch.dict(CC.COMPLETE_ATTRS, clear=True)
+def test__cmdln_args__empty_attrs(pytester):
+    """Fixture should return an empty config if COMPLETE_ATTRS is empty."""
+
+    patches = """
+        from tests.unit.conftest import MockConfig
+
+        @pytest.fixture(scope="session")
+        def request_config():
+            mock = MockConfig({"--users": 42, "--spawn_rate":.3})
+            return mock
+        """
+    pytester.makeconftest(
+        CONFTEST_TEMPLATE.format(fixture_name=FIXTURE_UNDER_TEST, patches=patches)
+    )
+    pytester.makepyfile(
+        PYFILE_TEMPLATE.format(
+            fixture_name=FIXTURE_UNDER_TEST,
+            validations=PYFILE_ASSERT_EMPTY_CONFIG.format(
+                fixture_name=FIXTURE_UNDER_TEST
+            ),
+        )
+    )
+
+    perform_fixture_test_with_optional_log_capture(pytester)
+
+
+@patch.dict(CC.COMPLETE_ATTRS, clear=True)
```

Comment:
what is this `@patch.dict` doing?

Reply:
Oh, patch.dict is cool, something I discovered just before pycon, but hadn't a chance to put it in the unit tests.

It lets you patch any dict - that is change the contents of the dict, but nothing else about the object. It also lets you control, do you want to just add some more values or clear the rest of the dict and only have the patched values. Used as a decorator, it is a context manager, meaning that it only patches the dict for the context of this single test.

In this case, the object was just a constant so I was mainly using its context manager capabilities to avoid side effects to other tests and having to keep track and reset the values all the time. However, sometimes it is very useful so that you don't have build out a mock that returns meaningful values for every method your code might access, when all you want is to change the contents of the underlying dict. A good example of this in our unit tests is patching os.environ. I wanted to set the env vars to a specific set of values, but didn't really want to mess around with mocking any other methods on this object (and also wanted the context manager portion).